etcd-client/src/watch/mod.rs

```diff
@@ -40,163 +40,619 @@
 //!
 //! ```
 
-use std::sync::Arc;
+use std::collections::{BTreeMap, HashMap};
+use std::sync::{Arc, Weak};
+use std::time::Duration;
 
-use async_trait::async_trait;
+use async_broadcast::{InactiveReceiver, Receiver as BroadcastRx, Sender as BroadcastTx};
+use async_std::channel::bounded;
+use crossbeam_queue::SegQueue;
 use futures::future::FutureExt;
-
 use futures::stream::StreamExt;
+use futures::{SinkExt, TryFutureExt};
+use grpcio::{ClientDuplexReceiver, StreamingCallSink, WriteFlags};
 use smol::channel::{unbounded, Receiver, Sender};
-use smol::stream::Stream;
+use smol::Task;
 
 pub use watch_impl::{EtcdWatchRequest, EtcdWatchResponse};
 
-use futures::prelude::*;
-
-use crate::lazy::{Lazy, Shutdown};
+use crate::lazy::Lazy;
 use crate::protos::kv;
-use crate::protos::rpc;
+use crate::protos::rpc::{WatchRequest, WatchResponse};
 use crate::protos::rpc_grpc::WatchClient;
-use crate::EtcdKeyValue;
 use crate::KeyRange;
 use crate::Result;
-use grpcio::WriteFlags;
+use crate::{EtcdError, EtcdKeyValue};
 
 /// Watch implementation mod.
 mod watch_impl;
 
+/// The timeout of waiting etcd response
+const WATCH_REQUEST_TIMEOUT_SEC: u64 = 2;
+
+/// Watch Id
+type WatchID = i64;
+
+/// Watch request message for watcher tasks to send to send loop
+type WatchRequestMsg = (EtcdWatchRequest, Sender<SingleWatchEventReceiver>);
+
+/// new watch request channel
+fn new_watch_request_chan() -> (Sender<WatchRequestMsg>, Receiver<WatchRequestMsg>) {
+    unbounded::<WatchRequestMsg>()
+}
+
+/// Cancel request message for watcher tasks to send to send loop
+type CancelRequestMsg = WatchID;
+
+/// new cancel request channel
+fn new_cancel_request_chan() -> (Sender<CancelRequestMsg>, Receiver<CancelRequestMsg>) {
+    unbounded::<CancelRequestMsg>()
+}
+
+/// Used for the send loop to receive the signal of a completed watch request from receive loop
+///  and handle the left watch request
+type WaitingWatchResponseMsg = ();
+
+/// new waiting watch response channel
+fn new_waiting_watch_response_chan() -> (
+    Sender<WaitingWatchResponseMsg>,
+    Receiver<WaitingWatchResponseMsg>,
+) {
+    bounded::<WaitingWatchResponseMsg>(1)
+}
+
+/// Shutdown message for send loop and receive loop async tasks to exit
+type ShutdownMsg = ();
+
+/// new shutdown channel
+fn new_shutdown_chan() -> (Sender<ShutdownMsg>, Receiver<ShutdownMsg>) {
+    unbounded::<ShutdownMsg>()
+}
+
+/// the recorded watch info in a lock
+#[derive(Default)]
+struct WatchedMap {
+    /// keyrange to watchid
+    keyrange_2_watchid: BTreeMap<KeyRange, WatchID>,
+    /// watchid to (sender to user, user receiver)
+    watchid_2_detail: HashMap<
+        WatchID,
+        (
+            BroadcastTx<EtcdWatchResponse>,
+            Weak<SingleWatchEventReceiverInner>,
+        ),
+    >,
```

Comment:
Why don't we just keep one map as they always appear at the same place?

Reply:
When a new task want to watch a key range, it will offer a keyrange, and the first map can find if there's watched same keyrange. The second map is for when we receive new watched event, an watch id is offered, and we can use watch id to directly get the BroadcastTx to broadcast event to all watchers.